text/0000-typed-context-injection.md

```diff
@@ -0,0 +1,833 @@
+
+```rust
+pub struct AudioBuffer { ... }
+
+impl AudioBuffer {
+    pub fn play_samples(&mut self, data: &[f32], volume: f32) { ... }
+}
+
+pub struct VirtualFileSystem { ... }
+
+impl VirtualFileSystem {
+    pub fn load_file(&mut self, path: &str) -> &[u8] { ... }
+}
+
+pub struct AudioCache { ... }
+
+impl AudioCache {
+    pub fn resolve_from_cache(&mut self, path: &str, loader: impl FnOnce() -> &[u8]) -> &[f32] { ... }
+}
+
+pub struct SoundVolumeOverrides { ... }
+
+impl SoundVolumeOverrides {
+    pub fn get_volume_override(path: &str) -> f32 { ... }
+}
+
+// Ideally, we could just write:
+cx.play_sound_at_path("cbat.ogg");
+
+// But, unfortunately, we must write:
+audio_buffer.play_sound_at_path(&mut virtual_fs, &mut audio_cache, &mut sound_volume_overrides, "cbat.ogg");
+```
+
+Yuck!
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Enter typed context injection. 
+
+Context injection is accomplished with two new standard library items: `Cx` and `AnyCx`. `Cx` is a variadic type parameterized by the references comprising a function's context. For example, to write a function taking `&mut System1` and `&System2`, you can just write:
+
+```rust
+fn consumer(cx: Cx<&mut System1, &System2>) { ... }
+```
+
+From there, you can fetch your components by either coercing the context into its comprising elements (which we'll henceforth refer to as *components*) or by using the `.extract::<T>()` and `.extract_mut::<T>()` utility methods, which perform the coercion internally but expose it in an occasionally-more-convenient method form.
+
+```rust
+fn consumer(cx: Cx<&mut System1, &System2>) {
+    let system_1: &mut System1 = cx;
+    let system_2: &System2 = cx;
+
+    // ...which is equivalent to:
+    let system_1 = cx.extract_mut::<System1>();
+    let system_2 = cx.extract::<System2>();
+}
+```
+
+To provide a `Cx` to a method, you can either create it with the `Cx::new` constructor or coerce a superset context into the target subset of itself.
+
+```rust
+fn caller_1(cx: Cx<&mut System1, &mut System2, &System3>) {
+    // `cx` has a mutable reference to `System1` and (at least) an immutable reference to `System2`
+    // so this is fine. The unnecessary `&System3` reference is completely ignored.
+    consumer(cx);
+}
+
+fn caller_2() {
+    ...
+
+    // We can give the context in the order it shows up in the target type...
+    consumer(Cx::new((&mut system_1, &system_2)));
+
+    // ...but we could also provide the context in any other order and let coercion
+    // turn e.g. a `Cx<&mut System2, &mut System1>` into a `Cx<&mut System1, &System2>`.
+    consumer(Cx::new((&mut system_2, &mut system_1)));
+
+    // Heck, we could even pass the function additional completely unrelated components and coercion
+    // would still have our back!
+    //
+    // Coerces `Cx<&mut SomeOtherSystem, &mut System2, &mut System1>` into
```

Comment:
Not really. In `.split`ting a context, you pass the entire context to the function. So, if you have an outstanding borrow on the context, it will be invalidated. In using borrow-aware coercion instead, borrow checking becomes much richer in a way that `.split` cannot emulate.

```rust
fn example(cx: Cx<&mut CommonSystem, &mut System1, &mut System2, &mut OtherSystem>) {
    let borrow_1: Cx<&CommonSystem, &mut System1> = cx;
    let borrow_2: Cx<&CommonSystem, &mut System2> = cx;

    // Both borrows can coexist.
    let _ = (borrow_1, borrow_2);

    // And you can get the values back once you're done!
    let borrow_3: Cx<&mut CommonSystem, &System1> = cx;

    // ...just so long as you don't use `borrow_1` or `borrow_2`.
    // let _ = (borrow_1, borrow_2);
}
```

And, yeah, you could maybe handle splits manually, but that really isn't convenient when you have a bunch of overlapping contexts that you need to recombine for one function call just to split them again for the next function call.